### ISSUE TITLE
Blazor Server side, child component not always updates view

### ISSUE THREAD

**Issue by Knudel** [aspnet/Blazor#1573]
I notices that a child component that used async entityframework not always got updated. So I tried to simulate the situation without Entity framework. 

I created a child component with this code:

```
@if(list == null)
{
    <div>
        Loading....
    </div>
}
else
{
    foreach(var i in list)
    {
        <div>
            @i
        </div>
    }
}

@functions
{
    protected List<string> list = null;

    protected override async Task OnInitAsync()
    {
        await Task.Run(async () =>
        {
            await Task.Delay(3000);

            list = new List<string>();
          
            list.Add("Some item 1");
            list.Add("Some item 2");
            list.Add("Some item 3");
            list.Add("Some item 4");
        });

        this.StateHasChanged();
    }
}
```

In de main page: index.cshtml insert A few child components like:

```
<Child></Child>
<Child></Child>
<Child></Child>
<Child></Child>
```

Open debugging with f5.
When the page has been loaded in chrome, reload with f5.

After a few time, a few components won't show but keep saying: loading....

**Comment by Andrzej-W** [comment]
Maybe related to #1223

**Comment by cores-system** [comment]
@Knudel, In asynchronous methods, you must call `StateHasChanged()`

**Comment by Knudel** [comment]
@cores-system I agree, but this doesn't help. My original code was a connection to a aql server with .net Entitycore. There I did call Statehaschanged().  I updated my code with StateHasChanges but it doesn't change.

Thank you for your help..

**Comment by danroth27** [comment]
@Andrzej-W I don't think calling `StateHasChanged` at the end of `OnInitAsync` should be necessary. I believe Blazor captures the returned task and calls `StateHasChanged` for you aftre the task completes.

@SteveSandersonMS I was able to reproduce this issue:

![image](https://user-images.githubusercontent.com/1874516/47187744-33521e00-d2ea-11e8-80f2-fb876a80dab7.png)

**Comment by Lupusa87** [comment]
I had same issue with different variations several times (it is same in SBB and CSB).
Last one was that when I select item there was not updating UI (selected item should change background color and in code notify parent about selection change for future actions like add, delete, update and etc).
I have child component and in parent's BuildRenderTree method I am dynamically rendering them, when I add new child I generate new unique ID.
On click event I call statehaschanged, in result BuildRenderTree is invoked again and is filling ID property with new unique values, when blazor calculates diffs it sees changed IDs and thinks that this changes should be applied to UI, removes all old html and adds new for new items.
Here main problem is that, blazor when calculates diffs sometimes thinks that there is no need to update html but actually it is.
This dirty way worked for me and I was able to force blazor to re-render UI.
I know it is not best solution because I need extra code and my app is doing extra job but for now I will use this before blazor renderer will be more accurate.
If it is possible I think there should be some method like StateHasChanged with will force blazor to re-render entire html on component level and don't calculate only diffs, for example UpdateEntireUI.
@SteveSandersonMS

**Comment by SteveSandersonMS** [comment]
Thanks for reporting this. It's fixed in #1672.
